optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/RuleMetadata.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import org.drools.model.Declaration;
+import org.drools.model.PatternDSL;
+
+public interface RuleMetadata<FactType> {
+
+    static LogicalRuleMetadata ofLogical(Declaration<LogicalTuple> variableDeclaration,
+            PatternDSL.PatternDef<LogicalTuple> pattern) {
+        return new LogicalRuleMetadata(variableDeclaration, pattern);
+    }
+
+    static <A> OriginalRuleMetadata<A> of(Declaration<A> variableDeclaration, PatternDSL.PatternDef<A> pattern) {
+        return new OriginalRuleMetadata<>(variableDeclaration, pattern);
+    }
+
+    Declaration<FactType> getVariableDeclaration();
+
+    PatternDSL.PatternDef<FactType> getPattern();
+}
```

Comment:
This interface (and its two implementors) exists so that we can enforce some type-safety within `Anchor`s - specifically, that the variable and its pattern are always of the same type. Code that keeps and treats these items separately has proven to be hard to understand.